Add Elastic search

maxlorenz opened the pull request:
<!--
(Thanks for sending a pull request! Please make sure you click the link above to view the contribution guidelines, then fill out the blanks below.)
-->
<!-- Add the issue number that is fixed by this PR (In the form Fixes #123) -->

Fixes #4875

#### Checklist

- [x] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia) and my PR follows them.
- [x] My branch is up-to-date with the Upstream `development` branch.
- [x] The unit tests pass locally with my changes <!-- use `nosetests tests/unittests` to run all the tests -->
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)
<!-- If an existing function does not have a docstring, please add one -->
- [x] All the functions created/modified in this PR contain relevant docstrings.

#### Short description of what this resolves:
Adds elasticsearch to the open event server. 

#### Changes proposed in this pull request:

- Adds elasticsearch `docker-compose` entry
- Adds ES config
- Adds first endpoint to search all events

schedutron commented on app/views/elastic_cron_helpers.py:
```diff
@@ -19,83 +31,38 @@ def cron_rebuild_events_elasticsearch():
+        searchable = SearchableEvent()
+        searchable.meta.id = event_id
+
+        if event_id:  # For safety
```
What is the need of this `if` block?

maxlorenz replied:
On create, the id is taken from the database and therefore can't be `None`. On delete however, it is possible that redis is out of sync so I double check the id here. It's not completely necessary though

maxlorenz commented on app/api/__init__.py:
```diff
@@ -508,3 +509,6 @@
 api.route(AdminSalesInvoicesList, 'admin_sales_invoices', '/admin/sales/invoices')
+
+# Search
+api.route(SearchEvent, 'search_event', '/search/event')
```
Okay, I will integrate that into the `/events` endpoint

maxlorenz replied:
I put it at `/events/query` now, e.g. `/events/query?name=eventname&location_name=location`

niranjan94 commented:
So, requirements are as follows.

- Accept a filter operation called as `search` as a part of the `filter` query param already being accepted by the `GET /events` endpoint.

	**Example:**
	
	```js
	filter: [
	  {
	    name : 'event',
	    op   : 'search',
	    val  : 'keyword-to-search-for-via-elasticsearch'
	  }
	]
	```
- Use ES to search for it
- Return the event response as usual. (As normal.) The full event response along with any other relationships & filtering/sorting that was requested. 

--

So, basically, this should just supplement what is already there as a part of `/events` endpoint.

niranjan94 commented:
> So basically we would have to search on ES and load the same IDs from Postgres afterwards? I'd prefer to implement full-text search on Postgres before going that route then, what do you think?

Yes. ES will search it's index. And give us the IDs. And we just quickly get the IDs from postgres and return the full response. And, no, we don't want to implement full-text on postgres. ES is meant for these kind of things. Also, we will be using ES to search across sessions too in the future. So, we need ES.

maxlorenz commented:
I added PG full-text search for events only. Will progress on ES but the solution for PG can be merged and shipped and changed in-place once ES is ready

niranjan94 commented:
> I added PG full-text search for events only. Will progress on ES but the solution for PG can be merged and shipped and changed in-place once ES is ready

@maxlorenz we don't want a PG full-text search. This PR/issue is for ES. And that is exactly what we want to have. So, adding a full-text here doesn't really make sense. So, please add ES instead as per the original requirement.

maxlorenz commented:
@niranjan94 the search is functional now, at the `/event?search-filter=[{some json}]` as described. `filter` cannot be used since the library would evaluate that to display only events with `search` properties (thus returning nothing every time)

mariobehling commented:
This is a big topic. Can we proceed with this as WIP and alter things later, please?

@maxlorenz Travis is not building atm.
